Adding multiple-URI support in `parse_uri` and `make_uri`

U65535F opened the pull request:
A rewritten version of #9756 that is almost similar to it.
The previous source got deleted somehow (apologies).

U65535F commented:
Tests are passing though.

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -15071,6 +15190,40 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
+  }
+  if (addresses.size() > 1)
+  {
+    error = "Multi-recipient URIs currently unsupported in this overload";
```
This is moreso a message for a developer, instead of a message for a user. The user (especially a non-technical one) will not have any idea what this (an overload) means. Consider rephrasing.

U65535F replied:
What should I rephrase it as? That is mistake of the developer right?

Tzadiko replied:
Im not sure if you'll ever hit this condition, unless there's a bug. Nonetheless I suggest you change it to something like "Multiple recipient addresses parsed out of transaction intended for one recipient.". You're moreso telling the user that there is a bug in your parse_uri moreso than there is an error that they themselves can fix. Hence, Im not sure how valuable this check even is. I want someone else to chime in.

Tzadiko replied:
Why was this marked as resolved without being addressed? Please only mark as resolved after you've pushed changes.

Tzadiko commented on src/wallet/wallet_rpc_server.cpp:
```diff
@@ -3142,6 +3142,22 @@ namespace tools
     return true;
   }
   //------------------------------------------------------------------------------------------------------------------------------
+  bool wallet_rpc_server::on_make_uri_v2(const wallet_rpc::COMMAND_RPC_MAKE_URI_V2::request& req, wallet_rpc::COMMAND_RPC_MAKE_URI_V2::response& res, epee::json_rpc::error& er, const connection_context *ctx)
```
I'm personally not a fan of the naming `_v2`. If a new developer comes and reads this, they will have no idea which one to use (v1 or v2). V2 may imply that V1 is no longer used, or that V2 is 'better' than V1. Consider renaming this to `on_make_uri_multiple` or something similar. You'll need to make this change to similar methods like `on_parse_uri_v2` and lines like these: https://github.com/monero-project/monero/pull/9830/files#diff-ffc587ad00e4573484c0430da9e42b7690f664de49df6f52b1a5be9b48833605R129

U65535F replied:
Yes. The developer doesn't know if he gets a muti recipient URI. So, the v1 function fails. v2 can process both URIs.  Take a look at old PR.

U65535F replied:
@Tzadiko

iamamyth replied:
As I recall, the version suffix exists because there was a stated intent to deprecate and eventually eliminate the prior endpoint, in favor of the new endpoint. If the plan changes to support both indefinitely, then I agree, v2 would be a poor naming choice.

U65535F replied:
Yes because the user never knows what type of uri is coming, it could be single recipient or multi. In case of multi uri, the legacy function fails. The newer functions work on both cases. If we continue on supporting both legacy and newer functions, it would introduce complexities in the code. First the user has to check if the uri is multi by calling the legacy function, then if it is, then call the newer function which handles the data in a slightly different way using vectors.

Tzadiko commented on src/wallet/wallet_rpc_server_commands_defs.h:
```diff
@@ -1829,7 +1829,7 @@ namespace wallet_rpc
     std::string address;
     std::string payment_id;
-    uint64_t amount;
+    std::uint64_t amount;
```
Why std::? I haven't seen this on anything other than MSVC, which isn't supported by Monero.

U65535F replied:
Suggested by SNeedleWoods at #9756

U65535F replied:
![image](https://github.com/user-attachments/assets/f96a617c-3de4-4230-bb86-9c069a9c06be)
@Tzadiko

Tzadiko replied:
In practice, many codebases use uint32_t without the std:: prefix for brevity, especially when working with C-compatible code or when the project includes <stdint.h>. They usually use std::uint32_t when they include <cstdint>. Just use uint32_t.

U65535F replied:
Alright 👍.

Tzadiko commented on tests/functional_tests/uri.py:
```diff
@@ -223,7 +231,92 @@ def test_monero_uri(self):
+        assert parsed.uri.tx_description == 'multi test'
+        assert parsed.uri.payment_id == ''
+
+        # build multi-recipient URI with three payments.
```
Can you make this into multiple tests? Youre testing quite a bit here, and each comment before the assert suggests something else is being tested. You can create a test fixture and use it to pass data that's common to each test to.

U65535F replied:
Please elaborate. I don't understand.

Tzadiko replied:
Sure, so you have six groupings of asserts in one test method (test_multi_uri). You have lines 252 to 261, lines 269 to 279, lines 290, lines 296, lines 310 to 311, and lines 316 to 319. Each of these groupings as a comment marked "Case: ...". So clearly there are 7 cases being tested. Each case should be its own test with a unique test name. Otherwise, if this massive test fails, it will be hard to determine which part failed (despite a line number being associated with the failure message). 

If some of these tests share data, build a test fixture that contains said data and use it accross multiple tests to avoid code duplication. More on test fixtures here: https://docs.pytest.org/en/6.2.x/fixture.html

U65535F replied:
OK. So you mean to split the `test_multi_uri` function into multiple functions, because it is vast and combines multiple test cases into one making it complex? 
Yes, those tests also share some variables. So you mean to pass the variables via test fixtures in order to avoid code redundancy. 

> Otherwise, if this massive test fails, it will be hard to determine which part failed (despite a line number being associated with the failure message).

It's easy, at least for me (as far I've seen) to determine which part fails, due to Python giving the line number. Not sure what you meant by `hard` but the suggestions seems to be good, helping the code to be more readable.

Tzadiko replied:
> So you mean to split the test_multi_uri function into multiple functions, because it is vast and combines multiple test cases into one making it complex?

Yes

> So you mean to pass the variables via test fixtures in order to avoid code redundancy.

Yes

> It's easy, at least for me (as far I've seen) to determine which part fails, due to Python giving the line number.

Let me share an example, I was looking at functional_tests_rpc.py yesterday. It calles a transfer test in transfer.py. transfer.py's `transfer` test is massive and failed on line number 177. Naturally I looked at the function call closest to this line failure to see what the issue was. I ended up spending 20 minutes looking at how this function call may have been affected by my changeset only to then notice the variable that was being asserted was from a call 20+ lines before this assert, that had several other asserts before it. So yes, while you may have the line number, it greatly reduces confusion and the need to navigate around the test when there is one thing under test per test, instead of many things.

Tzadiko replied:
Looks better now. Thanks

U65535F commented on src/wallet/wallet2.cpp:
```diff
@@ -14942,56 +14942,136 @@ std::string wallet2::decrypt_with_view_secret_key(const std::string &ciphertext,
 }
 //----------------------------------------------------------------------------------------------------
-std::string wallet2::make_uri(const std::string &address, const std::string &payment_id, uint64_t amount, const std::string &tx_description, const std::string &recipient_name, std::string &error) const
+std::string wallet2::custom_conver_to_url_format(const std::string &uri) const
```
I don't think that's any wrong there because there a function known as `epee::net_utils::conver_to_url_format` also exists.

Tzadiko replied:
Just because net_utils has a mispelled function doesn't mean you should double down on the misspelling.

U65535F replied:
Alright. I will fix it. I actually thought that `conver` was the real usage here. 
I wonder why net_utils hasn't fixed that yet, who knows.

U65535F commented on src/wallet/api/wallet.h:
```diff
@@ -214,7 +214,9 @@ class WalletImpl : public Wallet
     virtual void startRefresh() override;
     virtual void pauseRefresh() override;
     virtual bool parse_uri(const std::string &uri, std::string &address, std::string &payment_id, uint64_t &amount, std::string &tx_description, std::string &recipient_name, std::vector<std::string> &unknown_parameters, std::string &error) override;
+    bool parse_uri(const std::string &uri, std::vector<std::string> &addresses, std::vector<std::uint64_t> &amounts, std::vector<std::string> &recipient_names, std::string &payment_id, std::string &tx_description, std::vector<std::string> &unknown_parameters, std::string &error) override;
```
Yes, exactly.

U65535F commented on src/wallet/wallet2.cpp:
```diff
@@ -15036,31 +15129,57 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
+      boost::split(names_split, kv[1], boost::is_any_of(";"));
+      if (names_split.size() != addresses.size())
+      {
+        error = "Incorrect recipient_name count";
```
Can you name all of them please so I can fix them?

Tzadiko replied:
I think it's a bit of common sense. Review your changed error messages, anything that doesn't explicitly tell you what is wrong / is unactionable is a bit worthless. For example, `Incorrect tx_amount count` doesn't tell you what's wrong with the transaction count (that the sizes and amounts are not equal). Something like `tx_amount count of X is not equal to address count of Y` (fill in X and Y).

U65535F commented on src/wallet/api/wallet.h:
```diff
@@ -214,7 +214,9 @@ class WalletImpl : public Wallet
     virtual bool parse_uri(const std::string &uri, std::string &address, std::string &payment_id, uint64_t &amount, std::string &tx_description, std::string &recipient_name, std::vector<std::string> &unknown_parameters, std::string &error) override;
+    bool parse_uri(const std::string &uri, std::vector<std::string> &addresses, std::vector<std::uint64_t> &amounts, std::vector<std::string> &recipient_names, std::string &payment_id, std::string &tx_description, std::vector<std::string> &unknown_parameters, std::string &error) override;
     virtual std::string make_uri(const std::string &address, const std::string &payment_id, uint64_t amount, const std::string &tx_description, const std::string &recipient_name, std::string &error) const override;
+    std::string make_uri(const std::vector<std::string> &addresses, const std::vector<std::uint64_t> &amounts, const std::vector<std::string> &recipient_names, const std::string &payment_id, const std::string &tx_description, std::string &error) const override;
```
![image](https://github.com/user-attachments/assets/f96a617c-3de4-4230-bb86-9c069a9c06be)
SNeedlewoods (see #9756) said that they don't want to have "non-standard" types as arguments for methods in the Wallet API. FYI, we had a type for simplifying things as you said but later to the comments, I removed it and modified the code so it only accepts standard types.

U65535F replied:
@Tzadiko

U65535F commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -6532,25 +6532,71 @@ bool simple_wallet::transfer_main(const std::vector<std::string> &args_, bool ca
+        de.is_subaddress = info.is_subaddress;
+        de.is_integrated = info.has_payment_id;
+
+        if (info.has_payment_id || !payment_id_uri.empty())
```
I don't really know to be exact. But I think, yes it could be possible. Like wise, we have so many checks in the code which hardly hit.

U65535F commented:
@Tzadiko

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -14942,56 +14942,136 @@ std::string wallet2::decrypt_with_view_secret_key(const std::string &ciphertext,
+    {
+      addresses_str += ";";
+    }
+    addresses_str += address;
```
Why was this marked as resolved without being addressed?

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -15001,24 +15081,37 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
+  amounts.assign(addresses.size(), 0);
+  recipient_names.assign(addresses.size(), "");
+
+  if (ptr == NULL)
```
Why was this marked as resolved without being addressed?

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -15036,31 +15129,57 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
       }
       payment_id = kv[1];
+
+      if (payment_id.length() != 16 && payment_id.length() != 64)
```
Why was this marked as resolved without being addressed?

U65535F commented:
> > @Tzadiko
> 
> One tag is fine to prod me in case I missed the emails sent by Github. No need to tag me on each of my comments 👍

sorry, I thought I had to ping you in each conversation. Apologies.

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -15071,6 +15190,40 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
+  }
+  if (!amounts.empty() && amounts.size() != addresses.size())
+  {
+    error = "Incorrect amount count. Amount count must match address count. Zero may be use as a filler";
```
Why was this marked as resolved without being addressed? Please only mark as resolved after you've pushed changes.

Tzadiko commented:
You resolved a bunch of comments without them being addressed. I suggest we either come to a conclusion before they're resolved, or you make a change and push a commit. For example, after discussing I agree no further change is needed [here](https://github.com/monero-project/monero/pull/9830#discussion_r2001498757), so it can be resolved.

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -14974,13 +14974,13 @@ std::string wallet2::make_uri(const std::vector<std::string> &addresses, const s
   if (addresses.size() != amounts.size() || addresses.size() != recipient_names.size())
   {
-    error = "Mismatch between the number of addresses, amounts, and recipient names.";
+    error = std::string("The counts of addresses (" + addresses.size() + "), amounts (" + amounts.size() "), and recipient names (" + recipient_names.size() + ") do not match.");
```
This doesn't output what you think it does. This is doing `const char*` + `size_t` which is dangerous. You need to do:

`std::string{"The counts of addresses ("} + std::to_string(addresses.size())` ...

Or use boost::format / similar string formatting. See example as to why your code doesn't work: https://godbolt.org/z/1axqzrEdK

U65535F replied:
I realized this right after I pushed the commit. Fixed using boost::format.

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -15134,7 +15134,7 @@ bool wallet2::parse_uri(const std::string &uri, std::vector<std::string> &addres
       if (amounts_split.size() != addresses.size())
       {
-        error = "Incorrect tx_amount count";
+        error = std::string("tx_amount count of" + amounts_split.size() + "is not equal to address count of " + addresses.size());
```
Issue here too, same as previously mentioned.

Tzadiko commented on src/wallet/wallet2.cpp:
```diff
@@ -15191,14 +15193,9 @@ bool wallet2::parse_uri(const std::string &uri, std::vector<std::string> &addres
   if (!amounts.empty() && amounts.size() != addresses.size())
   {
-    error = "Incorrect amount count. Amount count must match address count. Zero may be use as a filler";
+    error = "Amount count does not match address count. Please use zero values where you don't want to specify the amounts.";
```
I think you mean "Please don't specify amounts if ..." instead of "Please use zero values". Although I'm not personally sure what not specifying amounts does.

U65535F replied:
Removed as no longer needed. Similar to the NA filter, it basically conveys that if the user cannot remove the addresses (in some case), then zeroing the amount that those addresses receive will help.

Tzadiko commented on tests/functional_tests/uri.py:
```diff
@@ -38,12 +39,18 @@
   from urllib.parse import quote as urllib_quote
 
 from framework.wallet import Wallet
+import pytest
```
In monero/tests/ there is a README.md file that lists the dependencies one needs to install to run the tests. I would add pytest there.

U65535F commented:
Can you have a look at `uri.py`, is there any modification required?
See line 211 and 156. Should I create a new function that calls all the multi URI test? Or just leave the `run_test` method intact?

U65535F commented:
> You resolved a bunch of comments without them being addressed. I suggest we either come to a conclusion before they're resolved, or you make a change and push a commit. For example, after discussing I agree no further change is needed [here](https://github.com/monero-project/monero/pull/9830#discussion_r2001498757), so it can be resolved.

Don't worry. Instead of letting you know every time (that seems spammy), I combined all of the suggestions into one commit and pushed them. Only time I will comment is when there is an issue.

U65535F commented:
I wonder why my test fixture isn't getting executed properly.
```
Traceback (most recent call last):
  File "/__w/monero/monero/tests/functional_tests/uri.py", line 369, in <module>
    URITest().run_test()
  File "/__w/monero/monero/tests/functional_tests/uri.py", line 48, in run_test
    self.test_multi_uri_two_payments()
  File "/__w/monero/monero/tests/functional_tests/uri.py", line 258, in test_multi_uri_two_payments
    wallet = multi_uri_common_variables["wallet"]
TypeError: 'URITest' object is not subscriptable
[TEST FAILED] uri
```

Tzadiko commented:
> I wonder why my test fixture isn't getting executed properly.
> 
> ```
> Traceback (most recent call last):
>   File "/__w/monero/monero/tests/functional_tests/uri.py", line 369, in <module>
>     URITest().run_test()
>   File "/__w/monero/monero/tests/functional_tests/uri.py", line 48, in run_test
>     self.test_multi_uri_two_payments()
>   File "/__w/monero/monero/tests/functional_tests/uri.py", line 258, in test_multi_uri_two_payments
>     wallet = multi_uri_common_variables["wallet"]
> TypeError: 'URITest' object is not subscriptable
> [TEST FAILED] uri
> ```

`multi_uri_common_variables()["wallet"]`

U65535F commented:
> > I wonder why my test fixture isn't getting executed properly.
> > 
> > ```
> > Traceback (most recent call last):
> >   File "/__w/monero/monero/tests/functional_tests/uri.py", line 369, in <module>
> >     URITest().run_test()
> >   File "/__w/monero/monero/tests/functional_tests/uri.py", line 48, in run_test
> >     self.test_multi_uri_two_payments()
> >   File "/__w/monero/monero/tests/functional_tests/uri.py", line 258, in test_multi_uri_two_payments
> >     wallet = multi_uri_common_variables["wallet"]
> > TypeError: 'URITest' object is not subscriptable
> > [TEST FAILED] uri
> > ```
> Call the fixture to get the returned dictionary, then access it: `multi_uri_common_variables()["wallet"]`

```
Traceback (most recent call last):
  File "/__w/monero/monero/tests/functional_tests/uri.py", line 369, in <module>
    URITest().run_test()
  File "/__w/monero/monero/tests/functional_tests/uri.py", line 48, in run_test
    self.test_multi_uri_two_payments()
  File "/__w/monero/monero/tests/functional_tests/uri.py", line 258, in test_multi_uri_two_payments
    wallet = multi_uri_common_variables()["wallet"]
TypeError: 'URITest' object is not callable
[TEST FAILED] uri
```
Now this.....

Tzadiko commented:
Hey, Im going to stop reviewing this. There's been a persistent theme of you tagging me 8 hours after having not heard from me, and it's getting on my nerves. I've seen you do this with other people as well. I already spoke with you about this, and you said you're not going to continue pinging me. For example, not even 8 hours after leaving my last comment, you've pinged me again - I'm a human being, not a GPT, I need to sleep, I just woke up and I don't work for you (and even if I did, this would still be unacceptable). Not to mention, you're pinging me for issues that can be solved with simple debugging skills on code that you've supposedly written yourself - I'm unsure why you to ping me with regards to setting up a Python test fixture, there's plenty of documentation and examples online.

Tzadiko commented:
I spent a bunch of time reviewing this and trying to help you. I wasn't going to leave another comment, until I saw your recent force-push. Instead of doing your part by doing some basic debugging to address a roadblock you encountered that was prompted by my comment, you reverted your code to a point beforewhich my comment existed for expedience. It's apparent that you don't care for improving the quality of your submission (which would in turn increase its chances of being accepted), which I was trying to help you improve, if it requires just a tiny bit of effort for you to figure out how to write a test fixture in Python.

I'm just as new as you, but this sort of lazy development doesn't have a place in this codebase, in my opinion. **I do not recommend anyone else review this PR, you'll be wasting your time.** OP will ignore it at the earliest sign of inconvenience if it means he gets his code crammed through faster.

U65535F commented:
> I spent a bunch of time reviewing this and trying to help you. I wasn't going to leave another comment, until I saw your recent force-push. Instead of doing your part by doing some basic debugging to address a roadblock you encountered that was prompted by my comment, you reverted your code to a point beforewhich my comment existed for expedience. It's apparent that you don't care for improving the quality of your submission (which would in turn increase its chances of being accepted), if it requires just a tiny bit of effort for you to figure out how to write a test fixture in Python.
> 
> I'm just as new as you, but this sort of lazy development doesn't have a place in this codebase, in my opinion. **I do not recommend anyone else review this PR, you'll be wasting your time.** OP will ignore it at the earliest sign of inconvenience if it means he gets his code crammed through faster.

Forgive. I have answers for your questions...
1. I'm new to programming especially into larger codebases and first of all, I'm new to contributing to FOSS. I started contributing and using GitHub properly this year. I started off with smaller repos (like the dead repos) and focusing on minor nits, then this is the first big PR I've done.
2. I've squashed the commits into one to make not mess this like the old PR. I have removed pytest as I think it's a overkill for just a small code which has different methods. I try hard to keep the implementation smaller and minimal as much as possible. I don't touch unnecessary areas of code, as devs criticized me in past for that.
3. At first, I really thought that GitHub was something similar to Discord, where a person only gets notified when they get pinged directly. (like @U65535F ). That was a mistake, I tried to rectify it to not pinging anyone altogether. Then I realized that a person also gets pinged when there's an activity in the PR (just realized). 
4. It was totally my problem as I wasn't very clear about my forward actions.
At the end, I still respected your contribution and it's still there in the code but it's squashed into one commit! The only thing removed is the pytest. Instead of pytest, I made a function which calls the core tests functions with the parameters as the common variables.
5. I thought that PRs used to get merged very quickly, like they get concluded in 2-3 days. But that wasn't true.

I'm extremely sorry for what I have done. I will try my best to be patient as much as possible and minimize the talk (as it notifies the users), additionally stop caring about the time PR's taking (give time for other users to respond; don't talk if they don't).

nahuhh commented:
You'll need to drop the merge commit and use rebase instead

iamamyth commented on tests/functional_tests/uri.py:
```diff
@@ -141,11 +143,17 @@ def test_monero_uri(self):
+        res = wallet.make_uri(address = address, recipient_name = utf8string[0], tx_description = utf8string[1], amount = 1000000000000, payment_id = '1' * 64)
+        assert res.uri == 'monero:' + address + '?tx_amount=1.000000000000&recipient_name=' + quoted_utf8string[0] + '&tx_description=' + quoted_utf8string[1] + '&tx_payment_id=' + '1' * 64
+        
+        # in case of standalone payment id removal
```
I'd rather you just eliminate this piece if the feature is still supported, it's an obvious one line change and the commented version only brings confusion.

U65535F replied:
I'm thinking whether to keep that part or not. Don't you think that it is kind of out of scope?

U65535F replied:
I think I'm going to leave standalone payment IDs deprecated in favor of integrated addresses.

U65535F commented:
@selsta I really appreciate if you decide to review this PR. Since the new release had been successfully planned and scheduled on Friday. I hope this time you will have some time. 

Major improvements from previous old review:
1. Introduce new RPC commands: `make_uri_v2` and `parse_uri_v2`: these will handle both single and multi recipient transactions. The old/legacy versions (i.e., `make_uri` and `parse_uri`) will be kept to fully ensure the backwards compatibility. (as suggested by iamamyth).
2. Remove most of the major unnecessary edits (there are couple of newlines there to tackle down).
3. Tests are passing successfully. Old tests have not been modified in any way.

U65535F commented:
@selsta Hello, can you please tell me when you'll review this. I will patiently wait till that date.

selsta commented:
The three points you have written above all sound good and were necessary to get this merged.

In file src/wallet/wallet_rpc_server_commands_defs.h, please update `WALLET_RPC_VERSION_MINOR` to indicate API changes.

U65535F commented:
@selsta Done with the changes. 👍

nahuhh commented:
> @selsta Done with the changes. 👍

@U65535F thanks @U65535F  @U65535F

FiatDemise commented:
Has this been tested with an integrated address? I have a use case where I'd like multi recipient URI to work with 1 integrated address and like 5 regular addresses.

U65535F commented:
> Has this been tested with an integrated address? I have a use case where I'd like multi recipient URI to work with 1 integrated address and like 5 regular addresses.

It technically should work but no test case has been added. I think I should add a test case for this.

selsta commented:
Since we do not allow external payment IDs anymore, it might make sense to completely remove them from the v2 API enpoints.

U65535F commented:
> Since we do not allow external payment IDs anymore, it might make sense to completely remove them from the v2 API enpoints. Only integrated addresses from now on. On the other hand, the URI spec still has tx_payment_id so we might have to support it or update the spec.
> 
> https://github.com/monero-project/monero/wiki/URI-Formatting

Instead of supporting external payment IDs, why don't we directly update the spec? I'm *unsure* because the legacy URI functions don't support external payment IDs either. 
https://github.com/monero-project/monero/blob/master/src/wallet/wallet2.cpp#L14963

selsta commented:
@U65535F I will get the specification updated to remove `tx_payment_id`, you can update the APIs.

selsta commented:
URI spec has been updated to remove `tx_payment_id` and support multiple addresses.

https://github.com/monero-project/monero/wiki/URI-Formatting

FiatDemise commented:
Is there a maximum number of recipients? Or a recommended number to stay under?

U65535F commented:
> Is there a maximum number of recipients? Or a recommended number to stay under?

No limits for processing the URI though. But I don't know about the backend or the part that processes those tx. The scope of this PR is only on parsing and crafting a URI from custom data like recipient names, tx description, addresses and amounts.

U65535F commented:
Changes done. Here's what I changed:
1. `make_uri_v2` and `parse_uri_v2` now don't support standalone/external payment id. If payment id is still present in the URI, `parse_uri_v2` will classify it as unknown parameter and push to the vector of such parameters. (`payment_id=id` format). The `v2` URI spec has also been updated to remove the payment id field from it.
2. However, the legacy versions still have the `payment_id` parameter and still support it clearly but they don't support multi-URI functionality though. This is done to fully ensure the backwards compatibility with older clients.
@selsta What do you think with the changes? Are they okay or is there anything I'm missing out? Any suggestions or any minor nits to fix? Though I will have to squash those commits into one after I'm done with the modifications.

U65535F commented:
@selsta What's your opinion on this?

selsta commented:
> parse_uri_v2 will classify it as unknown parameter and push to the vector of such parameters. (payment_id=id format).

Can you clarify this?

U65535F commented:
> > parse_uri_v2 will classify it as unknown parameter and push to the vector of such parameters. (payment_id=id format).
> 
> Can you clarify this?

Sure. If the url contains the payment_id field, it will be pushed to the unknown_parameter vector.

selsta commented:
No questions currently, I just have to go through all the code one more time and I didn't have time for it yet.

jsyk57 commented:
@selsta It's been almost 10 days. I will wait. Are you busy with the other PR reviews and the post-release work?

selsta commented:
I have time to go over it again today.

selsta commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -6583,34 +6608,20 @@ bool simple_wallet::transfer_main(const std::vector<std::string> &args_, bool ca
       {
-        fail_msg_writer() << tr("failed to set up payment id, though it was decoded correctly");
-        return false;
+          fail_msg_writer() << tr("failed to set up payment id");
```
Please check here and in other places in this file that the indentation is 2 spaces, if you change it back to 2 spaces the diff will get slightly smaller in some places.

selsta commented on src/wallet/wallet2.cpp:
```diff
@@ -15001,24 +15100,37 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
+  boost::split(arguments, params, boost::is_any_of("&"));
   std::set<std::string> have_arg;
-  for (const auto &arg: arguments)
+  for (const std::string &arg: arguments)
```
Is this neccessary?

selsta reviewed (commented):
Generally from my side I'm ok with the patch, only added 2 small comments. I don't usually review wallet changes, so others might have different opinions.

jsyk57 commented:
> I don't usually review wallet changes, so others might have different opinions.@selsta Would you approve this PR or will I have to wait for someone else to do it?

nahuhh commented:
monero is a collaboratively developed project. So, yes. You should be patient as others volunteer to review this PR.

This is not an immediate priority - there is not a release scheduled, and this doesnt fix anything. 

Just be patient. It could be a few weeks, or months, before this is reviewed by more people.
In the interim, you can move on to something else, or continue to make improvents here.

selsta commented on src/wallet/wallet2.cpp:
```diff
@@ -14942,56 +14942,155 @@ std::string wallet2::decrypt_with_view_secret_key(const std::string &ciphertext,
+  if (addresses.size() != amounts.size() || addresses.size() != recipient_names.size())
   {
-    error = "Standalone payment id deprecated, use integrated address instead";
+    error = (boost::format("The counts of addresses (%zu), amounts (%zu), and recipient names (%zu) do not match.") % addresses.size() % amounts.size() % recipient_names.size()).str();
```
```suggestion
    error = (boost::format("The counts of addresses (%1%), amounts (%2%), and recipient names (%3%) do not match.") % addresses.size() % amounts.size() % recipient_names.size()).str();
```

The current syntax appears incorrect. Please also update the other places that use `boost::format`.

https://www.boost.org/doc/libs/1_88_0/libs/format/doc/format.html

selsta commented on src/wallet/api/wallet.h:
```diff
@@ -214,7 +214,9 @@ class WalletImpl : public Wallet
     virtual void startRefresh() override;
     virtual void pauseRefresh() override;
     virtual bool parse_uri(const std::string &uri, std::string &address, std::string &payment_id, uint64_t &amount, std::string &tx_description, std::string &recipient_name, std::vector<std::string> &unknown_parameters, std::string &error) override;
+    bool parse_uri(const std::string &uri, std::vector<std::string> &addresses, std::vector<std::uint64_t> &amounts, std::vector<std::string> &recipient_names, std::string &tx_description, std::vector<std::string> &unknown_parameters, std::string &error) override;
```
```suggestion
    bool parse_uri(const std::string &uri, std::vector<std::string> &addresses, std::vector<uint64_t> &amounts, std::vector<std::string> &recipient_names, std::string &tx_description, std::vector<std::string> &unknown_parameters, std::string &error) override;
```

So that the signature matches with the existing one. Please do the same thing with `make_uri`.

jsyk57 replied:
thanks for introducing the problem, surprised it went overlooked till now. The entire code had `std::uint64_t` and `uint64_t` inconsistency. I've updated the code only to use `uint64_t` :)

selsta commented:
I will have one more round of suggestions tomorrow, but no major issues left.

selsta commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -6532,25 +6532,50 @@ bool simple_wallet::transfer_main(const std::vector<std::string> &args_, bool ca
-    bool has_uri = m_wallet->parse_uri(local_args[i], address_uri, payment_id_uri, amount, tx_description, recipient_name, unknown_parameters, error);
+    std::vector<std::string> addresses, recipient_names;
+    std::vector<uint64_t> amounts;
+    bool has_uri = m_wallet->parse_uri(local_args[i], addresses, amounts, recipient_names, tx_description, unknown_parameters, error);
```
Did you test if this correctly sets the description afterwards?

U65535F replied:
How am I supposed to test that?

U65535F replied:
> Compile it, and open moneo-wallet-cli and make a transaction with a URI containing a description and see if it sets the description.

I'm not really sure how I can see the description. I've tried googling and asked chatGPT but it was unsure.
This is what it says when I send a transaction to a wallet in testnet.
![image](https://github.com/user-attachments/assets/6d970d16-3d71-4351-8d1f-c30e3eaca103)
![image](https://github.com/user-attachments/assets/3114d118-3a48-45a9-adb7-2ebbac764f78)

U65535F replied:
@selsta I have modified the transfer_main code in simplewallet FYI.

U65535F replied:
@selsta

U65535F replied:
@selsta

selsta replied:
Modified the code but did you test what I wrote?

U65535F replied:
Yeah. I linked the image, see it.

selsta replied:
I was confused because it appears the whole URI gets saved in the description field instead of just the description, or am I misunderstanding the image?

U65535F replied:
> I was confused because it appears the whole URI gets saved in the description field instead of just the description, or am I misunderstanding the image?

I have no idea. That is what happens in the original app too iirc. I made so it achieves the same behavior as the original app.

U65535F replied:
@selsta Should we push the description to `Tx Note` field?

selsta replied:
I would say yes, it does not make sense as it is.

U65535F replied:
@selsta Alright but in the Destination field, should we display the full URI or just the address itself (the original app displays the full URI)? I wonder what happens to unknown parameters? Where are they stored?

nahuhh replied:
My 2c.

- Destination should only be the address 
- tx note should be the description
- the full uri should not be saved anywhere

selsta replied:
Agree with @nahuhh

U65535F replied:
![image](https://github.com/user-attachments/assets/c9863662-bd40-4bcd-8da9-e49f201e06cf)
Yes, it sets the description now (using `set_tx_note`). It doesn't store the full uri anywhere, now prints the address instead of the full uri like it used to do before. Note: the image has debug prints but I've removed them.

U65535F replied:
> Please show a tx using multi-uri

It is not afaik. It's the way how it formats (`show_transfers`). I'll link an image with multiuri wait.

U65535F replied:
@nahuhh Can you elaborate what I need to do?

nahuhh replied:
For a multiURI, showing the amount per address looks fine. 

for a single transfer, probably redundant to show. But not a dealbreaker if this was existing behavior

U65535F replied:
@selsta Can I resolve conversation now or are there any more things you can suggest?

selsta commented on src/wallet/wallet2.cpp:
```diff
@@ -15027,31 +15139,40 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
-      amount = 0;
-      if (!cryptonote::parse_amount(amount, kv[1]))
+      std::vector<std::string> amounts_split;
+      boost::split(amounts_split, kv[1], boost::is_any_of(";"));
```
Should `amounts_split.erase()` be added here like with addresses?

U65535F replied:
Yes. If there is a trailing semicolon, you will get an extra empty‐string token in the `amounts_split` vector which will cause the next if condition to fail. Thanks for pointing.

selsta reviewed (commented):
Can you add a test for multiple integrated addresses (should fail)?

U65535F commented:
@selsta Can you have a peek at simplewallet code and see if anything's wrong? Or if you can suggest something.

selsta commented:
Yes, will check tomorrow, I have to compile the PR and check how exactly it works now.

selsta commented:
I tested it in monero-wallet-cli and it looks good to me. I don't have any remaining suggestions at the moment.

selsta reviewed (approved):
I'm not sure if someone else is going to review this but from my side I'll add an approval.

sa8ab commented:
Is it possible to have multiple integrated addresses on one transaction?
Is it maximum 1 integrated address and rest sub address or primary addresses?

U65535F commented:
> Is it possible to have multiple integrated addresses on one transaction?
> Is it maximum 1 integrated address and rest sub address or primary addresses?

No multiple integrated addresses iirc.

jeffro256 commented on src/wallet/wallet_rpc_server.cpp:
```diff
@@ -3197,6 +3197,22 @@ namespace tools
+  {
+    if (!m_wallet) return not_open(er);
+    std::string error;
+    std::string uri = m_wallet->make_uri(req.addresses, req.amounts, req.recipient_names, req.tx_description, error);
```
This is why we shouldn't have `make_uri()`/`parse_uri()` be member functions of `wallet2`: you need to have initialized a `wallet2` instance for no reason.

U65535F replied:
alright, where would you like to see those free functions? Should I create a new source file or move the impl to some existing one?

jeffro256 commented on src/wallet/wallet_rpc_server.cpp:
```diff
@@ -3210,6 +3226,26 @@ namespace tools
+    std::vector<std::string> addresses;
+    std::vector<uint64_t> amounts;
+    std::vector<std::string> recipient_names;
+    if (!m_wallet->parse_uri(req.uri, addresses, amounts, recipient_names, res.uri.tx_description, res.unknown_parameters, error))
```
Same point here

jeffro256 commented on src/wallet/wallet2.cpp:
```diff
@@ -14933,56 +14933,155 @@ std::string wallet2::decrypt_with_view_secret_key(const std::string &ciphertext,
   }
 
-  if (amount > 0)
+  std::string uri = "monero:" + addresses_str;
```
I don't like the format of `<all addresses>?<all amounts>&<all messages>`, etc. Take a look at Bitcoin Unlimited's (node for Bcash) multiple address URI scheme: https://github.com/BitcoinUnlimited/BUIP/blob/master/086.md. The address, amount, and labels are next to each other for each output, which makes for more readable URIs. More importantly however, this allows the field for each output to remain undefined if desired. See this example for an actual use case: https://github.com/BitcoinUnlimited/BUIP/blob/master/086.md. Here you can see that the amount field for a tip output at a bar is left empty, which triggers the client code to prompt the user to enter a desired amount. Overall, this type of URI format is *way* more flexible and future-proof than cramming values into a list based on field type, assuming the same structure for every single output in a transaction.

selsta replied:
https://github.com/monero-project/monero/issues/7731#issuecomment-848102482

Here is some background on when it was decided on. I'm not sure if anyone actually implemented this.

jeffro256 replied:
Seems like @trasherdk and @benevanoff wanted a similar approach to what I am bringing up here: packing info per-output, instead of packing per-value-type.

jeffro256 replied:
I don't know if I'd worry much about the byte-length of the URIs unless it becomes a practical problem. Seems a bit like a premature optimization, IDK. QR codes can handle fairly large URLs.

FiatDemise replied:
> Seems like @trasherdk and @benevanoff wanted a similar approach to what I am bringing up here: packing info per-output, instead of packing per-value-type.

As far as I know, XMRChat is the first and only one to use this, and we don't have anything in prod yet. We're fine with whatever format you prefer. Just hope it can be done quickly as we'd like to implement this soon.

jeffro256 reviewed (changes requested):
I have some disagreements about the code structure, and the URI format itself.

iamamyth commented:
> Am I understanding correctly that this is the new format? Is there supposed to be a question mark at the beginning?
>
> monero:?output=address1;amount1;&output=address2;amount2;

Per my [comment](https://github.com/monero-project/monero/pull/9830#discussion_r2337738588), this looks wrong, as the original scheme specification (https://github.com/monero-project/monero/blob/master/docs/URI_SCHEME.md) contains no mention of an `output` param, and the single-output case should match the old format.

benevanoff commented:
@plowsof I was the original dev to start this implementation, about 3 years ago if memory serves correctly, but I had abandoned it because I wasnt getting much feedback on testing/code review. If this is something that people are now more interested in then I can just branch off the latest master and start from scratch. If this is the route we want to take then I will pop into IRC later in the day.

FiatDemise commented:
Hi @jeffro256 , in your proposed format at https://github.com/monero-project/monero/issues/7731#issuecomment-3239824513 , your example multi-recipient URI has "&amount=1XMR". Is the "XMR" needed, or can that part just be "&amount=1"?
